### PR TITLE
add BuildService

### PR DESCRIPTION
**Pull request by abe-winter**
## What changed
- Add BuildService with RPCs for starting and managing builds
## Reviewer notes
- We've marked this unstable until the implementation lands

### CODE REVIEW

File: proto/viam/app/build/v1/build.proto
Comment (michaellee1019) [inline proto/viam/app/build/v1/build.proto:34]: Shouldn't this be `repeated string build_id`? Based on the protos below, it feels like there will be a `build_id` generated per `platform`.

Or is JobInfo supposed to have `repeated string platform`

File: proto/viam/app/build/v1/build.proto
Comment (michaellee1019) [inline proto/viam/app/build/v1/build.proto:28]:
```suggestion
  repeated string platforms = 3;
```

File: proto/viam/app/build/v1/build.proto
Comment (michaellee1019) [inline proto/viam/app/build/v1/build.proto:45]: This might need to be a streaming API. How long are the logs? I forget our max length for a GRPC request, but given that we don't know the length and could get really long... Its probably best to just make it streaming now and avoid issues later.

File: proto/viam/app/build/v1/build.proto
Comment (michaellee1019) [inline proto/viam/app/build/v1/build.proto:76]: Whats the equivalent CLI command for this? I assume you'd have an optional `--build_id` param.

In our apis, we usually have separate Get and List apis. ListAPIs can have filters but usually not including filter using the primary key to get a single item. Usually have build_id as the only parameter to `GetJobRequest`. Any reason to stick with a single api?

**Review by michaellee1019 — CHANGES_REQUESTED** [review]
Nice first pass! Some general suggestions / questions here. Let me know if I'm off base because I haven't seen the whole system in code yet.

File: proto/viam/app/build/v1/build.proto
Comment (abe-winter) [reply proto/viam/app/build/v1/build.proto:37]: the model is 'one build contains multiple jobs'. each job represents one platform.

(and this maps idiomatically onto most of the vendors we've shopped)

File: proto/viam/app/build/v1/build.proto
Comment (abe-winter) [reply proto/viam/app/build/v1/build.proto:48]: getlogs api returns a stream

File: proto/viam/app/build/v1/build.proto
Comment (zaporter-work) [reply proto/viam/app/build/v1/build.proto]: Fixed. Good idea.

File: proto/viam/app/build/v1/build.proto
Comment (zaporter-work) [reply proto/viam/app/build/v1/build.proto:37]: The current impl is that each "build" kicks off multiple jobs (one per platform)

File: proto/viam/app/build/v1/build.proto
Comment (zaporter-work) [reply proto/viam/app/build/v1/build.proto:48]: It is a streaming API, that is what the comment says. Did I misunderstand your comment?

File: proto/viam/app/build/v1/build.proto
Comment (abe-winter) [reply proto/viam/app/build/v1/build.proto:79]: we're planning to flesh out CLI later

but in first release sticking to things that are strictly necessary to do this in CI

can be budged from that if your gut says we need a get

Comment (zaporter-work) [reply proto/viam/app/build/v1/build.proto:79]: This relates to the single buildid -> multiple platforms pattern (which might not be intuitive...)
This will return info about just the jobs that were kicked off for that build.
Ex: `viam module build list --id xyz123` (will just return info for the 3-4 architectures the module was built on)

File: proto/viam/app/build/v1/build.proto
Comment (michaellee1019) [reply proto/viam/app/build/v1/build.proto:37]: Makes sense now. And no plans for a job_id. Its the `build_id + platform` that makes a job unique?

File: proto/viam/app/build/v1/build.proto
Comment (zaporter-work) [reply proto/viam/app/build/v1/build.proto:48]: Good question. The reason I wrote it like this is primarily laziness (the build step will be ~10 bytes and the data string will be ~4MB so repeating the build step  a few times didn't really seem like a big deal). However, if this api is also confusing, that is bad. I will fix this

Comment (zaporter-work) [reply proto/viam/app/build/v1/build.proto:48]: > Does it work without oneof?

Yes, it just wastes a few bytes per message that shares the same build step

Comment (michaellee1019) [reply proto/viam/app/build/v1/build.proto:48]: Ok, I learned it isn't required today. I see other APIs that don't do it now too. I feel like the "convention" should be to use `oneof` but I think there are more usages without it than with. So we can pass on this.

**Comment by michaellee1019** [comment]
LGTM! Let me know if you want an approval now, or later (depends on when you want the protos to build).

**Review by michaellee1019 — APPROVED** [review]

**Review by michaellee1019 — APPROVED** [review]

**Pull request merged**
